File: frontend/src/types/wsMessages.ts
export interface NodeStats {
  node_id: string;
  label: string;
  node_type: string;   // "source" | "process" | "sink"
  queue_length: number;
  in_process: number;
  utilization: number;
  throughput: number;
  avg_cycle_time: number;
  total_completed: number;
}

export interface ResourceStats {
  resource_id: string;
  name: string;
  utilization: number;
  requests_queued: number;
}

export interface TickPayload {
  msg_type: "tick";
  sim_time: number;
  nodes: NodeStats[];
  resources: ResourceStats[];
  total_wip: number;
}

export interface SummaryPayload {
  msg_type: "summary";
  total_sim_time: number;
  nodes: NodeStats[];
  resources: ResourceStats[];
  total_throughput: number;
  sim_run_seconds: number;
}

export interface StatusPayload {
  msg_type: "status";
  state: "running" | "paused" | "stopped" | "warmup" | "idle";
}

export interface ErrorPayload {
  msg_type: "error";
  code: string;
  message: string;
}

export interface SimEvent {
  time: number;
  event_type: string;
  entity_id: number | null;
  node_id: string | null;
  node_label: string | null;
  resource_id: string | null;
  resource_label: string | null;
  resource_instance: number | null;
  details: string | null;
}

export interface EventLogPayload {
  msg_type: "event_log";
  events: SimEvent[];
  truncated: boolean;
}

export type WsMessage = TickPayload | SummaryPayload | StatusPayload | ErrorPayload | EventLogPayload;

File: backend/app/models/ws_messages.py
from __future__ import annotations
from typing import Any, Dict, List, Literal
from pydantic import BaseModel


class NodeStats(BaseModel):
    node_id: str
    label: str
    node_type: str = "process"   # "source" | "process" | "sink"
    queue_length: int = 0
    in_process: int = 0
    utilization: float = 0.0
    throughput: float = 0.0
    avg_cycle_time: float = 0.0
    total_completed: int = 0


class ResourceStats(BaseModel):
    resource_id: str
    name: str
    utilization: float = 0.0
    requests_queued: int = 0


class TickPayload(BaseModel):
    msg_type: Literal["tick"] = "tick"
    sim_time: float
    nodes: List[NodeStats]
    resources: List[ResourceStats]
    total_wip: int


class SummaryPayload(BaseModel):
    msg_type: Literal["summary"] = "summary"
    total_sim_time: float
    nodes: List[NodeStats]
    resources: List[ResourceStats]
    total_throughput: int
    sim_run_seconds: float


class StatusPayload(BaseModel):
    msg_type: Literal["status"] = "status"
    state: Literal["running", "paused", "stopped", "warmup", "idle"]


class ErrorPayload(BaseModel):
    msg_type: Literal["error"] = "error"
    code: str
    message: str


class EventLogPayload(BaseModel):
    msg_type: Literal["event_log"] = "event_log"
    events: List[Dict[str, Any]]   # serialized SimEvent dicts (via dataclasses.asdict)
    truncated: bool = False
